src/cli/term.rs

```diff
@@ -0,0 +1,160 @@
+    fn bg(&mut self, color: Color) -> io::Result<()> {
+        if !self.isatty() {
+            return Ok(());
+        }
+        self.color.set_bg(Some(color));
+        self.stream.set_color(&self.color)
+    }
+
+    fn attr(&mut self, attr: Attr) -> io::Result<()> {
+        if !self.isatty() {
+            return Ok(());
+        }
+        match attr {
+            Attr::Bold => self.color.set_bold(true),
+            Attr::ForegroundColor(color) => self.color.set_fg(Some(color)),
+        };
+        self.stream.set_color(&self.color)
+    }
+
+    fn reset(&mut self) -> io::Result<()> {
+        if !self.isatty() {
+            return Ok(());
+        }
+        self.stream.reset()
+    }
+
+    fn carriage_return(&mut self) -> io::Result<()> {
+        self.stream.write(b"\r").map(|_| ())
+    }
+}
+
+impl io::Write for AutomationFriendlyTerminal {
+    fn write(&mut self, buf: &[u8]) -> std::result::Result<usize, io::Error> {
+        self.stream.write(buf)
+    }
+
+    fn flush(&mut self) -> std::result::Result<(), io::Error> {
+        self.stream.flush()
+    }
+}
+
+impl AutomationFriendlyTerminal {
+    pub(crate) fn stdout() -> AutomationFriendlyTerminal {
+        let choice = if crate::utils::tty::stdout_isatty() {
+            ColorChoice::Auto
+        } else {
+            ColorChoice::Never
+        };
+        AutomationFriendlyTerminal {
+            stream: termcolor::StandardStream::stdout(choice),
+            color: ColorSpec::new(),
+        }
+    }
+
+    pub(crate) fn stderr() -> AutomationFriendlyTerminal {
+        let choice = if crate::utils::tty::stderr_isatty() {
+            ColorChoice::Auto
+        } else {
+            ColorChoice::Never
+        };
+        AutomationFriendlyTerminal {
+            stream: termcolor::StandardStream::stderr(choice),
+            color: ColorSpec::new(),
+        }
+    }
+}
+
+/// A cheaply clonable handle to a terminal.
+#[derive(Clone)]
+struct SharedTerminal {
+    inner: Arc<Mutex<Box<dyn Terminal + Send + Sync>>>,
+}
+
+impl io::Write for SharedTerminal {
+    fn write(&mut self, buf: &[u8]) -> std::result::Result<usize, io::Error> {
+        self.inner.lock().unwrap().write(buf)
+    }
+
+    fn flush(&mut self) -> std::result::Result<(), io::Error> {
+        self.inner.lock().unwrap().flush()
+    }
+}
+
+impl Terminal for SharedTerminal {
+    fn fg(&mut self, color: Color) -> io::Result<()> {
+        self.inner.lock().unwrap().fg(color)
+    }
+
+    fn bg(&mut self, color: Color) -> io::Result<()> {
+        self.inner.lock().unwrap().bg(color)
+    }
+
+    fn attr(&mut self, attr: Attr) -> io::Result<()> {
+        self.inner.lock().unwrap().attr(attr)
+    }
+
+    fn reset(&mut self) -> io::Result<()> {
+        self.inner.lock().unwrap().reset()
+    }
+
+    fn carriage_return(&mut self) -> io::Result<()> {
+        self.inner.lock().unwrap().carriage_return()
+    }
+}
+
+pub(crate) fn stdout() -> Box<dyn Terminal> {
+    Box::new(SharedTerminal {
```

Comment:
I've discovered that caching the stdout is quite difficult in this situation. As there are some tests running in the same process, using a once cell for caching is not a straightforward solution.

A possible solution could be to utilize `process().id()` to decide whether a new terminal needs to be created. However, this approach may involve certain lock options and mutability concerns. I am uncertain if this is a good solution since we aim for improved performance.

Reply:
Only cache the `isatty` to solve this problem. See https://github.com/rust-lang/rustup/pull/3351/commits/f76f961b407f731da9499446d8e7458fcbfb9e03

Reply:
so, process FileSource could answer isatty, to push that into the abstraction over env + fd's that is process().

Then caching that is local - it would have to drop the cache under some circumstances for correctness, but I don't think we ever play those games with fd 0/1/2 that would cause that to be needed.

That said, there are two things that raise concerns for me.

One is the removal of Writer from FileSource - that would then require Terminal to implement Writer itself - but it doesn't:

https://docs.rs/termcolor/latest/src/termcolor/lib.rs.html#443-445 shows that LockedStandardTerminal is the equivalent. 

For thread safety, and this is something Rust won't detect today (or perhaps it was and that is what you found hard last week) we *must* not use impl Write, but a locked Write (analogous to https://doc.rust-lang.org/std/io/struct.StdoutLock.html).

Failure to do that will lead to write striping from different threads.